zebra-scan/src/service/scan_task/commands.rs

```diff
@@ -0,0 +1,138 @@
+//! Types and method implementations for [`ScanTask`]
+
+use std::{
+    collections::HashMap,
+    sync::{
+        mpsc::{self, Receiver, TryRecvError},
+        Arc,
+    },
+};
+
+use color_eyre::{eyre::eyre, Report};
+use tokio::sync::oneshot;
+
+use zcash_primitives::{sapling::SaplingIvk, zip32::DiversifiableFullViewingKey};
+use zebra_chain::{block::Height, transaction::Transaction};
+use zebra_state::SaplingScanningKey;
+
+use super::ScanTask;
+
+#[derive(Debug)]
+/// Commands that can be sent to [`ScanTask`]
+pub enum ScanTaskCommand {
+    /// Start scanning for new viewing keys
+    RegisterKeys {
+        /// New keys to start scanning for
+        keys: HashMap<
+            SaplingScanningKey,
+            (Vec<DiversifiableFullViewingKey>, Vec<SaplingIvk>, Height),
+        >,
+    },
+
+    /// Stop scanning for deleted viewing keys
+    RemoveKeys {
+        /// Notify the caller once the key is removed (so the caller can wait before clearing results)
+        done_tx: oneshot::Sender<()>,
+
+        /// Key hashes that are to be removed
+        keys: Vec<String>,
+    },
+
+    /// Start sending results for key hashes to `result_sender`
+    SubscribeResults {
+        /// Sender for results
+        result_sender: mpsc::Sender<Arc<Transaction>>,
+
+        /// Key hashes to send the results of to result channel
+        key_hashes: Vec<()>,
+    },
+}
+
+impl ScanTask {
+    /// Accepts the scan task's `parsed_key` collection and a reference to the command channel receiver
+    ///
+    /// Processes messages in the scan task channel, updating `parsed_keys` if required.
+    ///
+    /// Returns newly registered keys for scanning.
+    pub fn process_messages(
+        cmd_receiver: &Receiver<ScanTaskCommand>,
+        parsed_keys: &mut HashMap<
+            SaplingScanningKey,
+            (Vec<DiversifiableFullViewingKey>, Vec<SaplingIvk>),
+        >,
+    ) -> Result<
+        HashMap<SaplingScanningKey, (Vec<DiversifiableFullViewingKey>, Vec<SaplingIvk>, Height)>,
+        Report,
+    > {
+        let mut new_keys = HashMap::new();
+
+        loop {
+            let cmd = match cmd_receiver.try_recv() {
+                Ok(cmd) => cmd,
+
+                Err(TryRecvError::Empty) => break,
+                Err(TryRecvError::Disconnected) => {
+                    // Return early if the sender has been dropped.
+                    return Err(eyre!("command channel disconnected"));
+                }
+            };
+
+            match cmd {
+                ScanTaskCommand::RegisterKeys { keys } => {
+                    let keys: Vec<_> = keys
+                        .into_iter()
+                        .filter(|(key, _)| parsed_keys.contains_key(key))
```

Comment:
Shouldn't be the filtering condition the other way around?

```suggestion
                        .filter(|(key, _)| !parsed_keys.contains_key(key))
```

Reply:
> Shouldn't be the filtering condition the other way around?

Yep. Whoops. Thank you for catching this, fixed in https://github.com/ZcashFoundation/zebra/pull/8222/commits/ba856c9379bea367ea8a7cdaf3b027c7fbb6f116.